TypeScript can't resolve non-relative paths in type imports

<!-- For feature requests, please use the following template: https://github.com/Shopify/polaris-react/issues/new?labels=Feature+request&template=FEATURE_REQUEST.md -->

## Issue summary

I'm getting a couple of errors when running the TypeScript compiler with Polaris 7.0.0:

```
[9:12:31 pm] File change detected. Starting incremental compilation...

node_modules/@shopify/polaris/build/ts/latest/src/components/OptionList/OptionList.d.ts:2:58 - error TS2307: Cannot find module 'types' or its corresponding type declarations.

2 import type { OptionDescriptor, SectionDescriptor } from 'types';
                                                           ~~~~~~~

node_modules/@shopify/polaris/build/ts/latest/src/types.d.ts:2:61 - error TS2307: Cannot find module 'components' or its corresponding type declarations.

2 import type { AvatarProps, IconProps, ThumbnailProps } from 'components';
                                                              ~~~~~~~~~~~~

[9:12:35 pm] Found 2 errors. Watching for file changes.
```

The two imports here are not using relative paths unlike most others, and when I update those two files locally to import `'../../types'` and `'./components'` (I think it was), the type check completes without error.

My tsconfig.json is:

```
{
  "compilerOptions": {
    "esModuleInterop": true,
    "jsx": "react",
    "moduleResolution": "node",
    "noEmit": true,
    "resolveJsonModule": true,
    "target": "ES2020"
  },
  "include": [
    "src/**/*"
  ]
}
```

## Specifications

- Are you using the React components? (Y/N): **Y**
- Polaris version number: **7.0.0**
- TypeScript version number: **4.3.5** (also tried with 4.4 and 4.2)
- Browser: **N/A**
- Device: **N/A**
- Operating System: Arch Linux

--- a/loom.config.ts
+++ b/loom.config.ts
@@ -53,7 +53,7 @@ function jestAdjustmentsPlugin() {
         // These do not work in rollup builds, so perhaps we shouldn't configure
         // them to work in jest tests either
         configuration.jestModuleNameMapper?.hook((moduleMapper) => {
-          moduleMapper['^(components|test-utilities|types|utilities)(.*)'] =
+          moduleMapper['^(components|test-utilities)(.*)'] =
             '<rootDir>/src/$1$2';
 
           return moduleMapper;

--- a/src/components/Banner/tests/Banner.test.tsx
+++ b/src/components/Banner/tests/Banner.test.tsx
@@ -7,7 +7,6 @@ import {
   DiamondAlertMajor,
 } from '@shopify/polaris-icons';
 import {mountWithApp} from 'test-utilities';
-import {BannerContext} from 'utilities/banner-context';
 import {
   Button,
   Heading,
@@ -17,6 +16,7 @@ import {
   UnstyledLink,
 } from 'components';
 
+import {BannerContext} from '../../../utilities/banner-context';
 import {WithinContentContext} from '../../../utilities/within-content-context';
 import {Banner, BannerHandles} from '../Banner';
 

--- a/src/components/Collapsible/tests/Collapsible.test.tsx
+++ b/src/components/Collapsible/tests/Collapsible.test.tsx
@@ -1,7 +1,7 @@
 import React, {useState, useCallback} from 'react';
 import {mountWithApp} from 'test-utilities';
-import {Tokens} from 'utilities/theme';
 
+import {Tokens} from '../../../utilities/theme';
 import {Collapsible, CollapsibleProps} from '../Collapsible';
 
 describe('<Collapsible />', () => {

--- a/src/components/Listbox/tests/Listbox.test.tsx
+++ b/src/components/Listbox/tests/Listbox.test.tsx
@@ -1,12 +1,12 @@
 import React, {useState} from 'react';
 import {mountWithApp} from 'test-utilities';
 import {animationFrame, timer} from '@shopify/jest-dom-mocks';
-import {Key} from 'types';
 import {
   mountWithComboboxListContext,
   mountWithListboxProvider,
 } from 'test-utilities/listbox';
 
+import {Key} from '../../../types';
 import {Button} from '../../Button';
 import {KeypressListener} from '../../KeypressListener';
 import {Scrollable} from '../../Scrollable';

--- a/src/components/OptionList/OptionList.tsx
+++ b/src/components/OptionList/OptionList.tsx
@@ -1,6 +1,10 @@
 import React, {useState, useCallback} from 'react';
-import type {Descriptor, OptionDescriptor, SectionDescriptor} from 'types';
 
+import type {
+  Descriptor,
+  OptionDescriptor,
+  SectionDescriptor,
+} from '../../types';
 import {isSection} from '../../utilities/options';
 import {arraysAreEqual} from '../../utilities/arrays';
 import {useUniqueId} from '../../utilities/unique-id';

--- a/src/components/PolarisTestProvider/tests/PolarisTestProvider.test.tsx
+++ b/src/components/PolarisTestProvider/tests/PolarisTestProvider.test.tsx
@@ -1,7 +1,7 @@
 import React, {StrictMode} from 'react';
 import {mount, mountWithApp} from 'test-utilities';
-import {MediaQueryContext, useMediaQuery} from 'utilities/media-query';
 
+import {MediaQueryContext, useMediaQuery} from '../../../utilities/media-query';
 import {PolarisTestProvider} from '../PolarisTestProvider';
 
 describe('PolarisTestProvider', () => {

--- a/src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx
+++ b/src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx
@@ -1,7 +1,7 @@
 import React from 'react';
 import {mountWithApp} from 'test-utilities';
-import {Key} from 'types';
 
+import {Key} from '../../../../../types';
 import {DualThumb, DualThumbProps} from '../DualThumb';
 
 describe('<DualThumb />', () => {

--- a/src/components/ResourceList/tests/ResourceList.test.tsx
+++ b/src/components/ResourceList/tests/ResourceList.test.tsx
@@ -10,8 +10,8 @@ import {
   EmptyState,
 } from 'components';
 import {mountWithApp} from 'test-utilities';
-import {SELECT_ALL_ITEMS} from 'utilities/resource-list';
 
+import {SELECT_ALL_ITEMS} from '../../../utilities/resource-list';
 import {BulkActions} from '../../BulkActions';
 import {CheckableButton} from '../../CheckableButton';
 import styles from '../ResourceList.scss';

--- a/src/types.ts
+++ b/src/types.ts
@@ -1,4 +1,4 @@
-import type {AvatarProps, IconProps, ThumbnailProps} from 'components';
+import type {AvatarProps, IconProps, ThumbnailProps} from './components';
 
 export interface OptionDescriptor {
   /** Value of the option */

--- a/src/utilities/options.ts
+++ b/src/utilities/options.ts
@@ -1,4 +1,4 @@
-import type {Descriptor, SectionDescriptor} from 'types';
+import type {Descriptor, SectionDescriptor} from '../types';
 
 export function isSection(arr: Descriptor[]): arr is SectionDescriptor[] {
   return (

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,7 +1,7 @@
 {
   "extends": "@shopify/typescript-configs/library",
   "compilerOptions": {
-    "baseUrl": "./src",
+    "baseUrl": "./",
     "rootDir": "./",
     "noEmit": false,
     "emitDeclarationOnly": true,
@@ -10,7 +10,13 @@
     "declarationMap": false,
     "isolatedModules": true,
     "importsNotUsedAsValues": "error",
-    "strictFunctionTypes": false
+    "strictFunctionTypes": false,
+    "paths": {
+      "components": ["src/components"],
+      "components/*": ["src/components/*"],
+      "test-utilities": ["src/test-utilities"],
+      "test-utilities/*": ["src/test-utilities/*"]
+    }
   },
   "include": ["./config/typescript", "./src", "./tests", "./playground"]
 }

